File: 65_Nim/ruby/nim.rb
puts "NIM".center(80)
puts"CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY".center(80)
puts "\n\n\n"

#210 DIM A(100),B(100,10),D(2)
$pileArray = Array.new[100]
$bArray = Array.new
$dArray = Array.new[2]
$winOption = 0 # take-last option
$numberOfPiles = 1
$c = 0
$e = 0
$f = 0
$g = 0
$h = 0

def displayTheRules
puts "THE GAME IS PLAYED WITH A NUMBER OF PILES OF OBJECTS."
puts "ANY NUMBER OF OBJECTS ARE REMOVED FROM ONE PILE BY YOU AND"
puts "THE MACHINE ALTERNATELY.  ON YOUR TURN, YOU MAY TAKE"
puts "ALL THE OBJECTS THAT REMAIN IN ANY PILE, BUT YOU MUST"
puts "TAKE AT LEAST ONE OBJECT, AND YOU MAY TAKE OBJECTS FROM"
puts "ONLY ONE PILE ON A SINGLE TURN.  YOU MUST SPECIFY WHETHER"
puts "WINNING IS DEFINED AS TAKING OR NOT TAKING THE LAST OBJECT,"
puts "THE NUMBER OF PILES IN THE GAME, AND HOW MANY OBJECTS ARE"
puts "ORIGINALLY IN EACH PILE.  EACH PILE MAY CONTAIN A"
puts "DIFFERENT NUMBER OF OBJECTS."
puts "THE MACHINE WILL SHOW ITS MOVE BY LISTING EACH PILE AND THE"
puts "NUMBER OF OBJECTS REMAINING IN THE PILES AFTER  EACH OF ITS"
puts "MOVES."
end

def sub1570
    $z=0
    for i in 1..$numberOfPiles do
        if $pileArray[i] != 0 then
            return
        end
        $z=1
        return
    end
end

def playAnother
    put "do you want to play another game";
    return gets.strip.ucase == "YES"
end
puts "THIS IS THE GAME OF NIM."
print "DO YOU WANT INSTRUCTIONS?"
240
wantInstructions = gets.strip.upcase
if wantInstructions == "YES" then
    displayTheRules
end
#250 IF Z$="NO" THEN 440
#260 IF Z$="no" THEN 440
#270 IF Z$="YES" THEN displayTheRules
#280 IF Z$="yes" THEN displayTheRules
#290 PRINT "PLEASE ANSWER YES OR NO"
#300 GOTO 240

def sub490 # get number of piles
    print "ENTER NUMBER OF PILES:"
    while $numberOfPiles < 0 && $numberOfPiles <= 100 do
        $numberOfPiles = gets.strip.to_i
    end
end

def getPileSizes
    puts "ENTER PILE SIZES:"
    for i in 1..$numberOfPiles do
        print i
        while true do
            $pileArray[i] = gets.strip.to_i
            if $pileArray[i] < 2000 && $pileArray[i] > 0 then
                break
            end
        end
    end
end

def sub440 # get win option
    puts ""
    $winOption = 0
    while $winOption != 1 && q != 2 do
        puts "ENTER WIN OPTION - 1 TO TAKE LAST, 2 TO AVOID LAST"
        $winOption = gets.strip.to_i
    end
end

puts "DO YOU WANT TO MOVE FIRST?";
#630 INPUT Q9$
moveFirst = ""
while moveFirst != "YES" && moveFirst != "NO" do
    moveFirst = gets.strip.upcase
    case moveFirst
        when "YES"
            yourMove
        when "NO"
            machineMove
    end
end

#640 IF Q9$="YES" THEN 1450
#650 IF Q9$="yes" THEN 1450
#660 IF Q9$="NO" THEN 700
#670 IF Q9$="no" THEN 700
#680 PRINT "PLEASE ANSWER YES OR NO."
#690 GOTO 630

def machineMove
    if $winOption == 1 then
        sub940 # take last
    end
    $c=0
    for i in 1..$numberOfPiles do
    if $pileArray[i] != 0 then 770
    $c=$c+1
    if $c == 3 then
        sub840
    end
        $dArray[$c-1]=i
    end
    end

    if $c == 2 then
        sub920
    end
    if $pileArray[$dArray[0]] > 1 then
        machineWins
    end
    machineLoses
end

def machineLoses
    puts "MACHINE LOSES"
# 810 GOTO playAnother
    if playAnother then
        sub440 # loop for another
    end
end

def machineWins
    puts "MACHINE WINS"
#    830 GOTO playAnother
    if playAnother then
        sub440 # loop for another
    end
end

def sub840
    $c=0
    for i in 1..$numberOfPiles do
    if $pileArray[i] > 1 then
        sub940
    end
    if $pileArray[i] == 0 then 890
        $c=$c+1
    end
    if $c/2 != ($c/2).to_i then
        machineLoses
    end
    sub940 # goto
    end
end

def sub920
    if $pileArray[$dArray[0]] == 1 then
        machineWins
    end
    if $pileArray[$dArray[1]] == 1 then
        machineWins
    end
end

def sub940
    for i in 1..$numberOfPiles do
        e=$pileArray[i]
        for j in 0..10 do
        $f = $e/2
        $bArray[i][j] = 2*($f-($f.to_i))
        $e = $f.to_i
        end
    end
end

#for j in 10..0 STEP -1 do
10..0.step(-1).each do|index|
    $c=0
    $h=0
    for i in 1..$numberOfPiles do
        if $bArray[i][index] != 0 then
            $c=$c+1
            if $pileArray[i] > $h then
                $h = $pileArray[i]
                $g = i
            end
        end
    end
end

if $c/2 != ($c/2).to_i then 1190
end
$e = rand($numberOfPiles).to_i
#if $pileArray[$e] == 0 then 1140

$f = rand($pileArray[$e]).to_i
$pileArray[$e] = $pileArray[$e]-$f
sub1380
$pileArray[$g]=0
for j in 0..10 do
$bArray[$g][index]=0
$c=0
for i in 1..$numberOfPiles do
    if $bArray[i][index] != 0 then
            $c=$c+1
        end
    end
$pileArray[$g]=$pileArray[$g]+2*($c/2-($c/2)).to_i*2^j
end
if $winOption == 1 then
    sub1380
end
$c=0
for i in 1..$numberOfPiles do
if $pileArray[i]>1 then
    sub1380
end
if $pileArray[i] != 0 then
    $c=$c+1
end
if $c/2 == ($c/2).to_i then
    sub1380
end
$pileArray[$g] == 1 -$pileArray[$g]

def sub1380
    puts "PILE  SIZE"
    for i in 1..$numberOfPiles do
        put i
        put $pileArray[i]
    end
    if $winOption == 2 then # avoid take-last option
        yourMove
    end
    sub1570
    if $z == 1 then
        machineWins
    end
end

def yourMove
    put "YOUR MOVE - PILE, NUMBER TO BE REMOVED"
#    1460 INPUT x,y
x = gets.strip.to_i
y = gets.strip.to_i
    if x > $numberOfPiles then yourMove
    if x < 1 then yourMove
    if x != INT(x) then yourMove
    if y > $pileArray[x] then yourMove
    if y < 1 then
        yourMove
    end
    if y != INT(y) then
        yourMove
    end

    $pileArray[x] = $pileArray[x]-y
    sub1570 # gosub
    if $z == 1 then
        machineLoses
    end
#    1560 GOTO 700
end

end
end
end
end
end

File: 00_Alternate_Languages/65_Nim/javascript/nim.js
// NIM
//
// Converted from BASIC to Javascript by Oscar Toledo G. (nanochess)
//

function print(str)
{
    document.getElementById("output").appendChild(document.createTextNode(str));
}

function input()
{
    var input_element;
    var input_str;

    return new Promise(function (resolve) {
                       input_element = document.createElement("INPUT");

                       print("? ");
                       input_element.setAttribute("type", "text");
                       input_element.setAttribute("length", "50");
                       document.getElementById("output").appendChild(input_element);
                       input_element.focus();
                       input_str = undefined;
                       input_element.addEventListener("keydown", function (event) {
                                                      if (event.keyCode == 13) {
                                                      input_str = input_element.value;
                                                      document.getElementById("output").removeChild(input_element);
                                                      print(input_str);
                                                      print("\n");
                                                      resolve(input_str);
                                                      }
                                                      });
                       });
}

function tab(space)
{
    var str = "";
    while (space-- > 0)
        str += " ";
    return str;
}

var a = [];
var b = [];
var d = [];

// Main program
async function main()
{
    print(tab(33) + "NIM\n");
    print(tab(15) + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY\n");
    print("\n");
    print("\n");
    print("\n");
    for (i = 1; i <= 100; i++) {
        a[i] = 0;
        b[i] = [];
        for (j = 0; j <= 10; j++)
            b[i][j] = 0;
    }
    d[0] = 0;
    d[1] = 0;
    d[2] = 0;
    print("DO YOU WANT INSTRUCTIONS");
    while (1) {
        str = await input();
        str = str.toUpperCase();
        if (str == "YES" || str == "NO")
            break;
        print("PLEASE ANSWER YES OR NO\n");
    }
    if (str == "YES") {
        print("THE GAME IS PLAYED WITH A NUMBER OF PILES OF OBJECTS.\n");
        print("ANY NUMBER OF OBJECTS ARE REMOVED FROM ONE PILE BY YOU AND\n");
        print("THE MACHINE ALTERNATELY.  ON YOUR TURN, YOU MAY TAKE\n");
        print("ALL THE OBJECTS THAT REMAIN IN ANY PILE, BUT YOU MUST\n");
        print("TAKE AT LEAST ONE OBJECT, AND YOU MAY TAKE OBJECTS FROM\n");
        print("ONLY ONE PILE ON A SINGLE TURN.  YOU MUST SPECIFY WHETHER\n");
        print("WINNING IS DEFINED AS TAKING OR NOT TAKING THE LAST OBJECT,\n");
        print("THE NUMBER OF PILES IN THE GAME, AND HOW MANY OBJECTS ARE\n");
        print("ORIGINALLY IN EACH PILE.  EACH PILE MAY CONTAIN A\n");
        print("DIFFERENT NUMBER OF OBJECTS.\n");
        print("THE MACHINE WILL SHOW ITS MOVE BY LISTING EACH PILE AND THE\n");
        print("NUMBER OF OBJECTS REMAINING IN THE PILES AFTER  EACH OF ITS\n");
        print("MOVES.\n");
    }
    while (1) {
        print("\n");
        while (1) {
            print("ENTER WIN OPTION - 1 TO TAKE LAST, 2 TO AVOID LAST");
            w = parseInt(await input());
            if (w == 1 || w == 2)
                break;
        }
        while (1) {
            print("ENTER NUMBER OF PILES");
            n = parseInt(await input());
            if (n >= 1 && n <= 100)
                break;
        }
        print("ENTER PILE SIZES\n");
        for (i = 1; i <= n; i++) {
            while (1) {
                print(i + " ");
                a[i] = parseInt(await input());
                if (a[i] >= 1 && a[i] <= 2000)
                    break;
            }
        }
        print("DO YOU WANT TO MOVE FIRST");
        while (1) {
            str = await input();
            str = str.toUpperCase();
            if (str == "YES" || str == "NO")
                break;
            print("PLEASE ANSWER YES OR NO.\n");
        }
        if (str == "YES")
            player_first = true;
        else
            player_first = false;
        while (1) {
            if (!player_first) {
                if (w != 1) {
                    c = 0;
                    for (i = 1; i <= n; i++) {
                        if (a[i] == 0)
                            continue;
                        c++;
                        if (c == 3)
                            break;
                        d[c] = i;
                    }
                    if (i > n) {
                        if (c == 2) {
                            if (a[d[1]] == 1 || a[d[2]] == 1) {
                                print("MACHINE WINS\n");
                                break;
                            }
                        } else {
                            if (a[d[1]] > 1)
                                print("MACHINE WINS\n");
                            else
                                print("MACHINE LOSES\n");
                            break;
                        }

                    } else {
                        c = 0;
                        for (i = 1; i <= n; i++) {
                            if (a[i] > 1)
                                break;
                            if (a[i] == 0)
                                continue;
                            c++;
                        }
                        if (i > n && c % 2) {
                            print("MACHINE LOSES\n");
                            break;
                        }
                    }
                }
                for (i = 1; i <= n; i++) {
                    e = a[i];
                    for (j = 0; j <= 10; j++) {
                        f = e / 2;
                        b[i][j] = 2 * (f - Math.floor(f));
                        e = Math.floor(f);
                    }
                }
                for (j = 10; j >= 0; j--) {
                    c = 0;
                    h = 0;
                    for (i = 1; i <= n; i++) {
                        if (b[i][j] == 0)
                            continue;
                        c++;
                        if (a[i] <= h)
                            continue;
                        h = a[i];
                        g = i;
                    }
                    if (c % 2)
                        break;
                }
                if (j < 0) {
                    do {
                        e = Math.floor(n * Math.random() + 1);
                    } while (a[e] == 0) ;
                    f = Math.floor(a[e] * Math.random() + 1);
                    a[e] -= f;
                } else {
                    a[g] = 0;
                    for (j = 0; j <= 10; j++) {
                        b[g][j] = 0;
                        c = 0;
                        for (i = 1; i <= n; i++) {
                            if (b[i][j] == 0)
                                continue;
                            c++;
                        }
                        a[g] = a[g] + (c % 2) * Math.pow(2, j);
                    }
                    if (w != 1) {
                        c = 0;
                        for (i = 1; i <= n; i++) {
                            if (a[i] > 1)
                                break;
                            if (a[i] == 0)
                                continue;
                            c++;
                        }
                        if (i > n && c % 2 == 0)
                            a[g] = 1 - a[g];
                    }
                }
                print("PILE  SIZE\n");
                for (i = 1; i <= n; i++)
                    print(" " + i + "  " + a[i] + "\n");
                if (w != 2) {
                    if (game_completed()) {
                        print("MACHINE WINS");
                        break;
                    }
                }
            } else {
                player_first = false;
            }
            while (1) {
                print("YOUR MOVE - PILE , NUMBER TO BE REMOVED");
                str = await input();
                x = parseInt(str);
                y = parseInt(str.substr(str.indexOf(",") + 1));
                if (x < 1 || x > n)
                    continue;
                if (y < 1 || y > a[x])
                    continue;
                break;
            }
            a[x] -= y;
            if (game_completed()) {
                print("MACHINE LOSES");
                break;
            }
        }
        print("DO YOU WANT TO PLAY ANOTHER GAME");
        while (1) {
            str = await input();
            str = str.toUpperCase();
            if (str == "YES" || str == "NO")
                break;
            print("PLEASE ANSWER YES OR NO.\n");
        }
        if (str == "NO")
            break;
    }
}

function game_completed()
{
    for (var i = 1; i <= n; i++) {
        if (a[i] != 0)
            return false;
    }
    return true;
}

main();
